altair/datasets/_reader.py

```diff
@@ -0,0 +1,555 @@
+    if reader := next(it, None):
+        return reader
+    raise AltairDatasetsError.from_priority(priority)
+
+
+@overload
+def _from_backend(name: _Polars, /) -> Reader[pl.DataFrame, pl.LazyFrame]: ...
+@overload
+def _from_backend(name: _PandasAny, /) -> Reader[pd.DataFrame, nw.LazyFrame]: ...
+@overload
+def _from_backend(name: _PyArrow, /) -> Reader[pa.Table, nw.LazyFrame]: ...
+
+
+# FIXME: The order this is defined in makes splitting the module complicated
+# - Can't use a classmethod, since some result in a subclass used
+def _from_backend(name: _Backend, /) -> Reader[Any, Any]:
+    """
+    Reader initialization dispatcher.
+
+    FIXME: Works, but defining these in mixed shape functions seems off.
+    """
+    if not _is_backend(name):
+        msg = f"Unknown backend {name!r}"
+        raise TypeError(msg)
+    implementation = _into_implementation(name)
+    if name == "polars":
+        rd, sc = _readimpl.pl_only()
+        return reader(rd, sc, name=name, implementation=implementation)
+    elif name == "pandas[pyarrow]":
+        return reader(_readimpl.pd_pyarrow(), name=name, implementation=implementation)
+    elif name == "pandas":
+        return reader(_readimpl.pd_only(), name=name, implementation=implementation)
+    elif name == "pyarrow":
+        return reader(_readimpl.pa_any(), name=name, implementation=implementation)
+
+
+def _is_backend(obj: Any) -> TypeIs[_Backend]:
+    return obj in {"polars", "pandas", "pandas[pyarrow]", "pyarrow"}
+
+
+def _is_err(obj: Any) -> TypeIs[type[AltairDatasetsError]]:
+    return obj is AltairDatasetsError
+
+
+def _into_constraints(
+    name: Dataset | LiteralString, suffix: Extension | None, /
+) -> Metadata:
+    """Transform args into a mapping to column names."""
+    m: Metadata = {}
+    if "." in name:
+        m["file_name"] = name
+    elif suffix is None:
+        m["dataset_name"] = name
+    elif suffix.startswith("."):
+        m = {"dataset_name": name, "suffix": suffix}
+    else:
+        from typing import get_args
+
+        from altair.datasets._typing import Extension
+
+        msg = (
+            f"Expected 'suffix' to be one of {get_args(Extension)!r},\n"
+            f"but got: {suffix!r}"
+        )
+        raise TypeError(msg)
+    return m
+
+
+def _into_implementation(
+    backend: _NwSupport | _PandasAny | Requirement, /
+) -> nw.Implementation:
+    primary = _import_guarded(backend)
+    mapping: Mapping[LiteralString, nw.Implementation] = {
+        "polars": nw.Implementation.POLARS,
+        "pandas": nw.Implementation.PANDAS,
+        "pyarrow": nw.Implementation.PYARROW,
+        "cudf": nw.Implementation.CUDF,
+        "dask": nw.Implementation.DASK,
+        "duckdb": nw.Implementation.DUCKDB,
+        "ibis": nw.Implementation.IBIS,
+        "pyspark": nw.Implementation.PYSPARK,
+    }
+    if impl := mapping.get(primary):
+        return impl
+    msg = f"Package {primary!r} is not supported by `narhwals`."
```

Comment:
```suggestion
    msg = f"Package {primary!r} is not supported by `narwhals`."
```

From @MarcoGorelli in (https://github.com/narwhals-dev/narwhals/pull/1734#discussion_r1938473775)